File: sc_custom/public/js/material_request.js
// ERPNext v16 (PR #53391) removed the supplier prompt that appeared when creating a
// Purchase Order from a Material Request, along with the server-side item filtering by
// default supplier. This restores the v15 behaviour for Purchase-type Material Requests:
// the "Create > Purchase Order" button asks for a supplier and the resulting PO contains
// only the items whose Item Default supplier matches (blank supplier => all items).

frappe.ui.form.on('Material Request', {
	custom_project: function (frm) {
		sc_apply_project_to_items(frm);
	},
	refresh: function (frm) {
		if (frm.doc.docstatus === 1 && frm.doc.material_request_type === 'Purchase') {
			// Replace the standard all-items "Purchase Order" button with the supplier-prompt one.
			// (Our refresh runs after erpnext's, which has already added the standard button.)
			frm.remove_custom_button('Purchase Order', 'Create');
			frm.add_custom_button(
				__('Purchase Order'),
				function () {
					sc_make_purchase_order_with_supplier(frm);
				},
				__('Create')
			);
		}
	},
});

// Show the item's default supplier / supplier part no as soon as it is picked, without
// waiting for a save (the server recomputes both columns on every save and submit).
frappe.ui.form.on('Material Request Item', {
	items_add: function (frm, cdt, cdn) {
		if (frm.doc.custom_project) {
			frappe.model.set_value(cdt, cdn, 'project', frm.doc.custom_project);
		}
	},
	item_code: function (frm, cdt, cdn) {
		let row = frappe.get_doc(cdt, cdn);

		if (!row.item_code || frm.doc.material_request_type !== 'Purchase') {
			frappe.model.set_value(cdt, cdn, 'custom_default_supplier', null);
			frappe.model.set_value(cdt, cdn, 'custom_supplier_part_no', null);
			return;
		}

		frappe.call({
			method: 'sc_custom.api.material_request.get_default_supplier_info',
			args: { item_code: row.item_code, company: frm.doc.company },
			callback: function (r) {
				// The row may have been deleted or repointed while the request was in flight.
				let current = locals[cdt] && locals[cdt][cdn];
				if (!r.message || !current || current.item_code !== row.item_code) return;
				frappe.model.set_value(cdt, cdn, 'custom_default_supplier', r.message.default_supplier);
				frappe.model.set_value(cdt, cdn, 'custom_supplier_part_no', r.message.supplier_part_no);
			},
		});
	},
});

// Copy the header Project into the items. If some rows already carry a different
// project the user chooses between overwriting them and filling only empty rows;
// closing the dialog leaves the items untouched.
function sc_apply_project_to_items(frm) {
	let project = frm.doc.custom_project;
	if (!project || !(frm.doc.items || []).length) return;

	let conflicting = frm.doc.items.filter((row) => row.project && row.project !== project);

	if (!conflicting.length) {
		sc_set_project_on_rows(frm, project, false);
		return;
	}

	let d = new frappe.ui.Dialog({
		title: __('Update Project in Items?'),
		fields: [
			{
				fieldtype: 'HTML',
				fieldname: 'message',
				options: `<p>${__(
					'{0} item row(s) already have a different Project. Overwrite them with {1}, or only fill rows without a Project?',
					[conflicting.length, `<b>${frappe.utils.escape_html(project)}</b>`]
				)}</p>`,
			},
		],
		primary_action_label: __('Overwrite All Rows'),
		primary_action() {
			d.hide();
			sc_set_project_on_rows(frm, project, true);
		},
		secondary_action_label: __('Only Fill Empty Rows'),
		secondary_action() {
			d.hide();
			sc_set_project_on_rows(frm, project, false);
		},
	});
	d.show();
}

function sc_set_project_on_rows(frm, project, overwrite) {
	let updated = 0;
	(frm.doc.items || []).forEach((row) => {
		if (row.project !== project && (overwrite || !row.project)) {
			frappe.model.set_value(row.doctype, row.name, 'project', project);
			updated++;
		}
	});
	if (updated) {
		frm.refresh_field('items');
		frappe.show_alert({
			message: __('Project set on {0} item row(s)', [updated]),
			indicator: 'green',
		});
	}
}

function sc_make_purchase_order_with_supplier(frm) {
	frappe.prompt(
		{
			label: __('For Default Supplier (Optional)'),
			fieldname: 'default_supplier',
			fieldtype: 'Link',
			options: 'Supplier',
			description: __(
				'Select a Supplier from the Default Suppliers of the items below. On selection, a Purchase Order will be made against items belonging to the selected Supplier only.'
			),
			get_query: function () {
				return {
					query: 'sc_custom.api.material_request.get_default_supplier_query',
					filters: { doc: frm.doc.name },
				};
			},
		},
		function (values) {
			frappe.model.open_mapped_doc({
				method: 'sc_custom.api.material_request.make_purchase_order_with_supplier',
				frm: frm,
				args: { default_supplier: values.default_supplier },
				run_link_triggers: true,
			});
		},
		__('Enter Supplier'),
		__('Create')
	);
}

File: sc_custom/doctype_events/material_request.py
"""Show each Material Request line's default supplier before a PO is created.

The restored "Create > Purchase Order" supplier flow (see api/material_request.py)
silently drops lines whose item has no matching Item Default supplier. These hooks
fill two read-only grid columns (custom_default_supplier, custom_supplier_part_no)
so users can spot incomplete item masters in advance, and warn about them on submit.
"""

import frappe
from frappe import _


def get_supplier_info_map(item_codes, company=None):
	"""Map item_code -> {default_supplier, supplier_part_no}.

	default_supplier comes from Item Default; the PO supplier filter matches it across
	all companies, so any filled row qualifies — a row of the given company just wins
	when several exist. supplier_part_no is the Item Supplier row of that supplier.
	"""
	item_codes = [ic for ic in set(item_codes or []) if ic]
	if not item_codes:
		return {}

	supplier_map = {}
	for d in frappe.get_all(
		"Item Default",
		filters={"parent": ["in", item_codes], "parenttype": "Item"},
		fields=["parent", "company", "default_supplier"],
	):
		if d.default_supplier and (d.parent not in supplier_map or d.company == company):
			supplier_map[d.parent] = d.default_supplier

	part_no_map = {
		(p.parent, p.supplier): p.supplier_part_no
		for p in frappe.get_all(
			"Item Supplier",
			filters={"parent": ["in", item_codes], "parenttype": "Item"},
			fields=["parent", "supplier", "supplier_part_no"],
		)
	}

	return {
		item_code: {
			"default_supplier": supplier,
			"supplier_part_no": part_no_map.get((item_code, supplier)),
		}
		for item_code, supplier in supplier_map.items()
	}


def set_default_supplier_info(doc, method=None):
	"""before_validate: refresh the info columns from the item master on every save.

	Runs in the submit cycle too, so the columns and the submit warning always reflect
	the current Item Default data even after the item master was fixed.
	"""
	info = {}
	if doc.material_request_type == "Purchase":
		info = get_supplier_info_map([d.item_code for d in doc.get("items", [])], doc.company)

	for d in doc.get("items", []):
		row_info = info.get(d.item_code) or {}
		d.custom_default_supplier = row_info.get("default_supplier")
		d.custom_supplier_part_no = row_info.get("supplier_part_no")


def warn_missing_default_supplier(doc, method=None):
	"""before_submit: non-blocking warning listing lines without a default supplier."""
	if doc.material_request_type != "Purchase":
		return

	missing = [d for d in doc.get("items", []) if d.item_code and not d.custom_default_supplier]
	if not missing:
		return

	rows = "<br>".join(_("Row {0}: {1}").format(d.idx, frappe.bold(d.item_code)) for d in missing)
	frappe.msgprint(
		_(
			"The following items have no Default Supplier in the Item master and will be"
			" skipped when a Purchase Order is created for a specific supplier:"
		)
		+ "<br>"
		+ rows,
		title=_("Missing Default Supplier"),
		indicator="orange",
	)
